sopht/simulator/immersed_body/cosserat_rod/cosserat_rod_forcing_grids.py

```diff
@@ -321,20 +326,112 @@ def __init__(
         ).astype(int)
         # If there are less than 1 point then set it equal to 1 since we will place it on the element center.
         self.surface_grid_points[np.where(self.surface_grid_points < 3)[0]] = 1
+
+        # Include cap grid points in end element surface grid points
+        if self.with_cap:
+            end_elem_grid_points_radius_ratio = []
+            end_elem_surface_grid_points = []
+            for rod_end_idx in [0, -1]:
+                rod_end_radius = self.cosserat_rod.radius[rod_end_idx]
+                rod_end_surface_grid_points = self.surface_grid_points[rod_end_idx]
+                if rod_end_surface_grid_points > 1:
+                    # we can still fit points within the surface grid circle
+                    # use max(arc-length of angular discretization, rod element length)
+                    # as reference length for radial discretization
+                    grid_angular_spacing = 2.0 * np.pi / rod_end_surface_grid_points
+                    reference_length = max(
+                        rod_end_radius * grid_angular_spacing,
+                        np.amax(self.cosserat_rod.lengths),
+                    )
```

Comment:
```suggestion
                    reference_length = rod_end_radius * grid_angular_spacing
```
also, it would be more intuitive to call it `end_elem_surface_grid_spacing` rather than `reference_length`

Reply:
Yes you are right. I thought I renamed it to that, or somewhere along those lines but I guess I didn't. I'll rename it and push it together with the refactoring you suggested in the comment below.

Reply:
btw, is it okay for us to only use the angular grid spacing as reference for end element (radial) grid spacing? Or do you see potential issues arising with comparing also the rod axial discretization length? @bhosale2

Reply:
I think since the outermost surface grid is fixed whether the cap is present or not, it's intuitive to use the local information of grid spacing on that element to compute the end cap grids. We can revisit this point later if any weird case arises, that requires considering rod lengths.

Reply:
Fixed in [a8ad220](https://github.com/SophT-Team/SophT/pull/121/commits/a8ad220cef0a4b6d88d60473cfe907e20f1b3766)